gadget/gadget.go

```diff
@@ -551,6 +562,93 @@ type VolumeUpdate struct {
 	Preserve []string       `yaml:"preserve" json:"preserve"`
 }
 
+// VolumeAssignment is an optional set of volume-to-disk assignments
+// that can be specified. This is a nice way of reusing the same gadget
+// for multiple devices, or a way to map multiple volumes to the same disk
+// for cases like eMMC. Each assignment in this structure refers to a volume
+// and a device path (/dev/disk/** for now).
+type VolumeAssignment struct {
+	Name        string                       `yaml:"assignment-name"`
+	Assignments map[string]*DeviceAssignment `yaml:"assignment"`
+}
+
+func (va *VolumeAssignment) validate(volumes map[string]*Volume) error {
+	if len(va.Assignments) == 0 {
+		return fmt.Errorf("no assignments specified")
+	}
+
+	for name, asv := range va.Assignments {
+		if v := volumes[name]; v == nil {
+			return fmt.Errorf("volume %q is mentioned in assignment but has not been defined", name)
+		}
+		if err := asv.validate(); err != nil {
+			return fmt.Errorf("%q: %v", name, err)
+		}
+	}
+	return nil
+}
+
+// DeviceAssignment is the device data for each volume assignment. Currently
+// just keeps the device the volume should be assigned to.
+type DeviceAssignment struct {
+	Device string `yaml:"device"`
+}
+
+func (da *DeviceAssignment) validate() error {
+	// TODO: Revise this in the future to maybe limit it even further
+	// as not all sub categories inside /dev/disk can be really useful.
+	if !strings.HasPrefix(da.Device, "/dev/disk/") {
+		return fmt.Errorf("unsupported device path %q, for now only paths under /dev/disk are valid", da.Device)
+	}
+	return nil
+}
+
+func areAssignmentsMatchingCurrentDevice(assignments map[string]*DeviceAssignment) bool {
+	for _, va := range assignments {
+		if _, err := os.Stat(path.Join(dirs.GlobalRootDir, va.Device)); err != nil {
+			// XXX: expect this to mean no path, consider actually
+			// ensuring the error is not-exists
+			logger.Debugf("failed to stat device %s: %v", va.Device, err)
+			return false
+		}
+	}
+	return true
+}
+
+// Indirection here for mocking
+var VolumesForCurrentDeviceAssignment = volumesForCurrentDeviceAssignmentImpl
+
+// volumesForCurrentDeviceAssignmentImpl does a best effort match of the volume-assignments
+// against the current device. We find the first assignment that has all device paths matching
```

Comment:
It looks like the implementation contradicts this comment, this is returning an error if there is more than one match. Which does not match (pun intended) the spec, which says that the first match should be picked. It could be possible to have more than one match in fact, we simply prioritize and get the first one.

Reply:
updated!